File: NurSamplePC_vs2008_VB/NurSamplePC_VB/Connection.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Text
Imports System.Windows.Forms
Imports NurApiDotNet

Public Partial Class Connection
	Inherits UserControl
	''' <summary>
	''' The NurApi handle
	''' </summary>
	Private hNur As NurApi = Nothing

	Private disableEvents As Boolean = False

	Private Class connectionComboItem
		Public DevPath As String
		Public Name As String
		Public Port As Integer
		Public Sub New(devPath As String, name As String, port As Integer)
			Me.DevPath = devPath
			Me.Name = name
			Me.Port = port
		End Sub
		Public Overrides Function ToString() As String
			Return Name
		End Function
	End Class

	''' <summary>
	''' Initializes a new instance of the <see cref="Connection"/> class.
	''' </summary>
	Public Sub New()
		InitializeComponent()
	End Sub

	Private Sub Connection_Load(sender As Object, e As EventArgs)
		usbCombo_Click(Me, EventArgs.Empty)
		serialCombo_Click(Me, EventArgs.Empty)
		'useLatestRadioBox.Checked = true;
		useUsbAutoRadioBox.Checked = True
	End Sub

	''' <summary>
	''' Sets the NurApi.
	''' </summary>
	''' <param name="hNur">The handle of NurApi.</param>
	Public Sub SetNurApi(hNur As NurApi)
		Try
			Me.hNur = hNur

			' Set event handlers for NurApi
			AddHandler hNur.DisconnectedEvent, New EventHandler(Of NurApi.NurEventArgs)(AddressOf hNur_DisconnectedEvent)
			AddHandler hNur.ConnectedEvent, New EventHandler(Of NurApi.NurEventArgs)(AddressOf hNur_ConnectedEvent)

			' Update the status of the connection
			If hNur.IsConnected() Then
				hNur_ConnectedEvent(hNur, Nothing)
			Else
				hNur_DisconnectedEvent(hNur, Nothing)
				conStatusDesc.Text = "No connection"
			End If
		Catch ex As NurApiException
			MessageBox.Show(ex.ToString(), Program.appName, MessageBoxButtons.OK, MessageBoxIcon.[Error])
		End Try
	End Sub

	''' <summary>
	''' Handles the DisconnectedEvent event of the NUR module.
	''' </summary>
	''' <param name="sender">The source of the event.</param>
	''' <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
	Private Sub hNur_DisconnectedEvent(sender As Object, e As NurApi.NurEventArgs)
		Dim hNur As NurApi = TryCast(sender, NurApi)
		conStatusDesc.Text = "Disconnected"
		UpdateButtons()
	End Sub

	''' <summary>
	''' Handles the ConnectedEvent event of the NUR module.
	''' </summary>
	''' <param name="sender">The source of the event.</param>
	''' <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
	Private Sub hNur_ConnectedEvent(sender As Object, e As NurApi.NurEventArgs)
		Dim hNur As NurApi = TryCast(sender, NurApi)
		conStatusDesc.Text = "Connected"
		UpdateButtons()
		Try
			Dim ethConfig As NurApi.EthConfig = hNur.GetEthConfig()
			tcpipAddr.Text = IpToString(ethConfig.ip)
			tcpipPort.Value = ethConfig.serverPort
				' Ignore this error
		Catch generatedExceptionName As NurApiException
		End Try
	End Sub

	Private Sub updateControls_CheckedChanged(sender As Object, e As EventArgs)
		If disableEvents Then
			Return
		End If

		usbCombo.Enabled = useUsbRadioBox.Checked
		serialCombo.Enabled = useSerialRadioBox.Checked
		tcpipAddr.Enabled = useTcpipRadioBox.Checked
		tcpipPort.Enabled = useTcpipRadioBox.Checked

		If hNur IsNot Nothing AndAlso useUsbAutoRadioBox Is sender Then
			hNur.SetUsbAutoConnect(useUsbAutoRadioBox.Checked)
		End If

		UpdateButtons()
	End Sub

	Private Sub usbCombo_Click(sender As Object, e As EventArgs)
		If hNur IsNot Nothing Then
			Dim ports As List(Of NurApi.UsbDevice) = NurApi.EnumerateUsbDevices()

			Dim currentSelection As Integer = usbCombo.SelectedIndex
			usbCombo.Items.Clear()
			For Each dev As NurApi.UsbDevice In ports
				usbCombo.Items.Add(New connectionComboItem(dev.devPath, dev.friendlyName, 0))
			Next
			If 0 <= currentSelection AndAlso currentSelection < usbCombo.Items.Count Then
				usbCombo.SelectedIndex = currentSelection
			Else
				usbCombo.SelectedIndex = usbCombo.Items.Count - 1
			End If
		End If
	End Sub

	Private Sub serialCombo_Click(sender As Object, e As EventArgs)
		If hNur IsNot Nothing Then
			Dim ports As List(Of NurApi.ComPort) = NurApi.EnumerateComPorts()

			Dim currentSelection As Integer = serialCombo.SelectedIndex
			serialCombo.Items.Clear()
			For Each port As NurApi.ComPort In ports
				serialCombo.Items.Add(New connectionComboItem("", port.friendlyName, port.port))
			Next
			If 0 <= currentSelection AndAlso currentSelection < serialCombo.Items.Count Then
				serialCombo.SelectedIndex = currentSelection
			Else
				serialCombo.SelectedIndex = serialCombo.Items.Count - 1
			End If
		End If
	End Sub

	Private Sub connectBtn_Click(sender As Object, e As EventArgs)
		Try
			If hNur.IsConnected() Then
				hNur.Disconnect()
			Else
				If useLatestRadioBox.Checked Then
					hNur.Connect()
				ElseIf useUsbAutoRadioBox.Checked Then
					hNur.SetUsbAutoConnect(False)
					hNur.SetUsbAutoConnect(True)
				ElseIf useUsbRadioBox.Checked Then
					Dim comboItem As connectionComboItem = TryCast(usbCombo.SelectedItem, connectionComboItem)
					hNur.ConnectUsb(comboItem.DevPath)
				ElseIf useSerialRadioBox.Checked Then
					Dim comboItem As connectionComboItem = TryCast(serialCombo.SelectedItem, connectionComboItem)
					hNur.ConnectSerialPort(comboItem.Port)
				ElseIf useTcpipRadioBox.Checked Then
					hNur.ConnectSocket(tcpipAddr.Text, CInt(Math.Truncate(tcpipPort.Value)))
				End If
			End If
		Catch ex As NurApiException
			MessageBox.Show(ex.ToString(), Program.appName, MessageBoxButtons.OK, MessageBoxIcon.[Error])
		End Try
	End Sub

	Private Shared Function IpToString(ip As Byte()) As String
		If ip Is Nothing Then
			Return System.Net.IPAddress.None.ToString()
		End If
		Dim ipAddress As New System.Net.IPAddress(ip)
		Return ipAddress.ToString()
	End Function

	Private Sub UpdateButtons()
		If hNur Is Nothing Then
			Return
		End If

		If hNur.IsConnected() Then
			connectBtn.Text = "DISCONNECT"
		Else
			connectBtn.Text = "CONNECT"
		End If
	End Sub
End Class

File: NurSamplePC_vs2008_CS/NurSamplePC/Connection.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using NurApiDotNet;

namespace NurSample
{
    public partial class Connection : UserControl
    {
        /// <summary>
        /// The NurApi handle
        /// </summary>
        NurApi hNur = null;

        bool disableEvents = false;

        class connectionComboItem
        {
            public string DevPath;
            public string Name;
            public int Port;
            public connectionComboItem(string devPath, string name, int port)
            {
                this.DevPath = devPath;
                this.Name = name;
                this.Port = port;
            }
            public override string ToString() { return Name; }
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Connection"/> class.
        /// </summary>
        public Connection()
        {
            InitializeComponent();
        }

        private void Connection_Load(object sender, EventArgs e)
        {
            socketAddressTextBox.Text = Properties.Settings.Default.connectionSocketAddress;
            socketPortNumericUpDown.Value = Properties.Settings.Default.connectionSocketPort;
            usbCombo_Click(this, EventArgs.Empty);
            serialCombo_Click(this, EventArgs.Empty);
            // Update RadioBoxes
            useUsbRadioBox.Checked = Properties.Settings.Default.connectionUsb;
            useUsbAutoRadioBox.Checked = Properties.Settings.Default.connectionUsbAuto;
            useSerialRadioBox.Checked = Properties.Settings.Default.connectionSerial;
            useSocketRadioBox.Checked = Properties.Settings.Default.connectionSocket;
        }

        /// <summary>
        /// Sets the NurApi.
        /// </summary>
        /// <param name="hNur">The handle of NurApi.</param>
        public void SetNurApi(NurApi hNur)
        {
            try
            {
                this.hNur = hNur;

                // Set event handlers for NurApi
                hNur.DisconnectedEvent += new EventHandler<NurApi.NurEventArgs>(hNur_DisconnectedEvent);
                hNur.ConnectedEvent += new EventHandler<NurApi.NurEventArgs>(hNur_ConnectedEvent);

                // Update the status of the connection
                if (hNur.IsConnected())
                    hNur_ConnectedEvent(hNur, null);
                else
                {
                    hNur_DisconnectedEvent(hNur, null);
                    conStatusDesc.Text = "No connection";
                }
            }
            catch (NurApiException ex)
            {
                MessageBox.Show(ex.ToString(), Program.appName, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        /// <summary>
        /// Handles the DisconnectedEvent event of the NUR module.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
        private void hNur_DisconnectedEvent(object sender, NurApi.NurEventArgs e)
        {
            NurApi hNur = sender as NurApi;
            conStatusDesc.Text = "Disconnected";
            UpdateButtons();
        }

        /// <summary>
        /// Handles the ConnectedEvent event of the NUR module.
        /// </summary>
        /// <param name="sender">The source of the event.</param>
        /// <param name="e">The <see cref="NurApi.NurEventArgs" /> instance containing the event data.</param>
        private void hNur_ConnectedEvent(object sender, NurApi.NurEventArgs e)
        {
            NurApi hNur = sender as NurApi;
            conStatusDesc.Text = "Connected";
            UpdateButtons();
            // Update default settings
            Properties.Settings.Default.connectionUsb = useUsbRadioBox.Checked;
            Properties.Settings.Default.connectionUsbAuto = useUsbAutoRadioBox.Checked;
            Properties.Settings.Default.connectionSerial = useSerialRadioBox.Checked;
            Properties.Settings.Default.connectionSocket = useSocketRadioBox.Checked;
            if (useSocketRadioBox.Checked)
            {
                Properties.Settings.Default.connectionSocketAddress = socketAddressTextBox.Text;
                Properties.Settings.Default.connectionSocketPort = (int)socketPortNumericUpDown.Value;
            }
            // Save default settings
            Properties.Settings.Default.Save();
        }

        private void updateControls_CheckedChanged(object sender, EventArgs e)
        {
            if (disableEvents)
                return;

            usbCombo.Enabled = useUsbRadioBox.Checked;
            serialCombo.Enabled = useSerialRadioBox.Checked;
            socketAddressTextBox.Enabled = useSocketRadioBox.Checked;
            socketPortNumericUpDown.Enabled = useSocketRadioBox.Checked;

            if (hNur != null && useUsbAutoRadioBox == sender)
            {
                hNur.SetUsbAutoConnect(useUsbAutoRadioBox.Checked);
            }

            UpdateButtons();
        }

        private void usbCombo_Click(object sender, EventArgs e)
        {
            if (hNur != null)
            {
                List<NurApi.UsbDevice> ports = NurApi.EnumerateUsbDevices();

                int currentSelection = usbCombo.SelectedIndex;
                usbCombo.Items.Clear();
                foreach (NurApi.UsbDevice dev in ports)
                {
                    usbCombo.Items.Add(new connectionComboItem(dev.devPath, dev.friendlyName, 0));
                }
                if (0 <= currentSelection && currentSelection < usbCombo.Items.Count)
                    usbCombo.SelectedIndex = currentSelection;
                else
                    usbCombo.SelectedIndex = usbCombo.Items.Count - 1;
            }
        }

        private void serialCombo_Click(object sender, EventArgs e)
        {
            if (hNur != null)
            {
                List<NurApi.ComPort> ports = NurApi.EnumerateComPorts();

                int currentSelection = serialCombo.SelectedIndex;
                serialCombo.Items.Clear();
                foreach (NurApi.ComPort port in ports)
                {
                    serialCombo.Items.Add(new connectionComboItem("", port.friendlyName, port.port));
                }
                if (0 <= currentSelection && currentSelection < serialCombo.Items.Count)
                    serialCombo.SelectedIndex = currentSelection;
                else
                    serialCombo.SelectedIndex = serialCombo.Items.Count - 1;
            }
        }

        private void connectBtn_Click(object sender, EventArgs e)
        {
            try
            {
                if (hNur.IsConnected())
                {
                    hNur.Disconnect();
                }
                else
                {
                    if (useLatestRadioBox.Checked)
                    {
                        hNur.Connect();
                    }
                    else if (useUsbAutoRadioBox.Checked)
                    {
                        hNur.SetUsbAutoConnect(false);
                        hNur.SetUsbAutoConnect(true);
                    }
                    else if (useUsbRadioBox.Checked)
                    {
                        connectionComboItem comboItem = usbCombo.SelectedItem as connectionComboItem;
                        hNur.ConnectUsb(comboItem.DevPath);
                    }
                    else if (useSerialRadioBox.Checked)
                    {
                        connectionComboItem comboItem = serialCombo.SelectedItem as connectionComboItem;
                        hNur.ConnectSerialPort(comboItem.Port);
                    }
                    else if (useSocketRadioBox.Checked)
                    {
                        hNur.ConnectSocket(socketAddressTextBox.Text, (int)socketPortNumericUpDown.Value);
                    }
                }
            }
            catch (NurApiException ex)
            {
                MessageBox.Show(ex.ToString(), Program.appName, MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        private void UpdateButtons()
        {
            if (hNur == null)
                return;

            if (hNur.IsConnected())
            {
                connectBtn.Text = "DISCONNECT";
            }
            else
            {
                connectBtn.Text = "CONNECT";
            }
        }
    }
}
